File: kuantumprojesiPython/kuantum.py
import uuid
import random
from abc import ABC, abstractmethod

# CUSTOM EXCEPTION
class KuantumCokusuException(Exception):
    def __init__(self, id):
        super().__init__(f"KUANTUM ÇÖKÜŞÜ! Nesne patladı: {id}")

# ABSTRACT BASE CLASS (ABC kullanımı)
class KuantumNesnesi(ABC):
    def __init__(self, id, stab, tehlike):
        self.__id = id  # Private attribute (name mangling)
        self._stabilite = stab
        self.tehlike = tehlike

    # Getter (ID sadece okunabilir)
    @property
    def id(self):
        return self.__id

    # Property ile Encapsulation (0-100 sınırlaması)
    @property
    def stabilite(self):
        return self._stabilite

    @stabilite.setter
    def stabilite(self, v):
        self._stabilite = max(0, min(100, v))

    # Abstract metot (zorunlu override)
    @abstractmethod
    def analiz_et(self):
        pass

    def durum_bilgisi(self):
        return f"[{self.id[:8]}...] Stabilite: {self.stabilite:.1f}% | Tehlike: {self.tehlike}"

# INTERFACE benzeri sınıf (Duck Typing)
class IKritik:
    @abstractmethod
    def acil_sogutma(self):
        pass

# CONCRETE CLASS 1: Güvenli veri
class VeriPaketi(KuantumNesnesi):
    def __init__(self, id):
        super().__init__(id, 100, 1)

    def analiz_et(self):
        self.stabilite -= 5
        print("Veri içeriği okundu.")

# CONCRETE CLASS 2: Karanlık madde (Multiple Inheritance)
class KaranlikMadde(KuantumNesnesi, IKritik):
    def __init__(self, id):
        super().__init__(id, 100, 7)

    def analiz_et(self):
        self.stabilite -= 15
        print("Karanlık madde analiz edildi.")

    def acil_sogutma(self):
        self.stabilite = min(100, self.stabilite + 50)
        print("Soğutma aktif! +50 stabilite")

# CONCRETE CLASS 3: Anti madde (Multiple Inheritance)
class AntiMadde(KuantumNesnesi, IKritik):
    def __init__(self, id):
        super().__init__(id, 100, 10)

    def analiz_et(self):
        print("Evrenin dokusu titriyor")
        self.stabilite -= 25

    def acil_sogutma(self):
        self.stabilite = min(100, self.stabilite + 50)
        print(" ACİL SOĞUTMA! +50 stabilite")

# MAIN LOOP
envanter = []

try:
    while True:
        print(" KUANTUM AMBARI KONTROL PANELİ")
        print("1. Yeni Nesne Ekle")
        print("2. Envanteri Listele")
        print("3. Nesneyi Analiz Et")
        print("4. Acil Durum Soğutması")
        print("5. Çıkış")

        s = input("\nSeçiminiz: ").strip()

        if s == "1":
            # Polimorfizm: Rastgele somut sınıf
            t = random.randint(0, 2)
            yeni_id = str(uuid.uuid4())

            if t == 0:
                nesne = VeriPaketi(yeni_id)
            elif t == 1:
                nesne = KaranlikMadde(yeni_id)
            else:
                nesne = AntiMadde(yeni_id)

            envanter.append(nesne)
            print(f"✓ {nesne.__class__.__name__} oluşturuldu.")
        
        elif s == "2":
            print(f"\n--- ENVANTER ({len(envanter)} nesne) ---")
            for x in envanter:
                print(x.durum_bilgisi())

        elif s == "3":
            arama_id = input("Nesne ID (ilk 8 karakter): ").strip()
            
            nesne = next((x for x in envanter if x.id.startswith(arama_id)), None)
            
            if nesne:
                nesne.analiz_et()
                
                if nesne.stabilite <= 0:
                    raise KuantumCokusuException(nesne.id)
            else:
                print(" Nesne bulunamadı!")

        elif s == "4":
            arama_id = input("Nesne ID (ilk 8 karakter): ").strip()
            
            nesne = next((x for x in envanter if x.id.startswith(arama_id)), None)
            
            if nesne:
                # Type Checking: isinstance ile IKritik kontrolü
                if isinstance(nesne, IKritik):
                    nesne.acil_sogutma()
                else:
                    print(" Bu nesne soğutulamaz! (IKritik değil)")
            else:
                print(" Nesne bulunamadı!")

        elif s == "5":
            print("Sistem kapatılıyor...")
            break

except KuantumCokusuException as e:
    print("\n" + "=" * 50)
    print(" SİSTEM ÇÖKTÜ! TAHLİYE BAŞLATILIYOR... 🚨")
    print(str(e))
    print("=" * 50)
except KeyboardInterrupt:
    print("\n\nSistem zorla sonlandırıldı.")

File: kuantumprojesiJs/kuantum.js
const readline = require("readline");
const { v4: uuidv4 } = require("uuid");

// CUSTOM EXCEPTION
class KuantumCokusuException extends Error {
    constructor(id) {
        super(`KUANTUM ÇÖKÜŞÜ! Nesne patladı: ${id}`);
        this.name = "KuantumCokusuException";
    }
}

// ABSTRACT BASE CLASS (JavaScript'te abstract simülasyonu)
class KuantumNesnesi {
    #id;  // Private field (ES2022)

    constructor(id, stab, tehlike) {
        if (new.target === KuantumNesnesi) {
            throw new Error("KuantumNesnesi soyut bir sınıftır!");
        }
        this.#id = id;
        this._stabilite = stab;
        this.tehlike = tehlike;
    }

    // Getter (ID sadece okunabilir)
    get id() {
        return this.#id;
    }

    // Encapsulation: Property ile 0-100 sınırlaması
    get stabilite() {
        return this._stabilite;
    }

    set stabilite(v) {
        this._stabilite = Math.max(0, Math.min(100, v));
    }

    // Abstract metot (override zorunlu)
    analizEt() {
        throw new Error("analizEt() metodu override edilmelidir!");
    }

    durumBilgisi() {
        return `[${this.id.substring(0, 8)}...] Stabilite: ${this.stabilite.toFixed(1)}% | Tehlike: ${this.tehlike}`;
    }
}

// INTERFACE simülasyonu (Symbol ile marker interface)
const IKritik = Symbol("IKritik");

// CONCRETE CLASS 1: Güvenli veri
class VeriPaketi extends KuantumNesnesi {
    constructor(id) {
        super(id, 100, 1);
    }

    analizEt() {
        this.stabilite -= 5;
        console.log(" Veri içeriği okundu.");
    }
}

// CONCRETE CLASS 2: Karanlık madde (IKritik marker)
class KaranlikMadde extends KuantumNesnesi {
    constructor(id) {
        super(id, 100, 7);
        this[IKritik] = true;  // Interface marker
    }

    analizEt() {
        this.stabilite -= 15;
        console.log(" Karanlık madde analiz edildi.");
    }

    acilDurumSogutmasi() {
        this.stabilite = Math.min(100, this.stabilite + 50);
        console.log(" Soğutma aktif! +50 stabilite");
    }
}

// CONCRETE CLASS 3: Anti madde (IKritik marker)
class AntiMadde extends KuantumNesnesi {
    constructor(id) {
        super(id, 100, 10);
        this[IKritik] = true;  // Interface marker
    }

    analizEt() {
        console.log("Evrenin dokusu titriyor...");
        this.stabilite -= 25;
    }

    acilDurumSogutmasi() {
        this.stabilite = Math.min(100, this.stabilite + 50);
        console.log(" ACİL SOĞUTMA! +50 stabilite");
    }
}

// Readline interface
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function soru(prompt) {
    return new Promise(resolve => rl.question(prompt, resolve));
}

// MAIN LOOP
(async () => {
    const envanter = [];

    try {
        while (true) {
            console.log("KUANTUM AMBARI KONTROL PANELİ");
            console.log("1. Yeni Nesne Ekle");
            console.log("2. Envanteri Listele");
            console.log("3. Nesneyi Analiz Et");
            console.log("4. Acil Durum Soğutması");
            console.log("5. Çıkış");

            const s = await soru("\nSeçiminiz: ");

            if (s === "1") {
                // Polimorfizm: Rastgele somut sınıf
                const t = Math.floor(Math.random() * 3);
                const yeniId = uuidv4();

                let nesne;
                if (t === 0) nesne = new VeriPaketi(yeniId);
                else if (t === 1) nesne = new KaranlikMadde(yeniId);
                else nesne = new AntiMadde(yeniId);

                envanter.push(nesne);
                console.log(` ${nesne.constructor.name} eklendi.`);
            }
            else if (s === "2") {
                console.log(`\n ENVANTER (${envanter.length} nesne)`);
                envanter.forEach(x => console.log(x.durumBilgisi()));
            }
            else if (s === "3") {
                const aramaId = await soru("Nesne ID (ilk 8 karakter): ");
                const nesne = envanter.find(x => x.id.startsWith(aramaId.trim()));

                if (nesne) {
                    nesne.analizEt();

                    if (nesne.stabilite <= 0) {
                        throw new KuantumCokusuException(nesne.id);
                    }
                } else {
                    console.log(" Nesne bulunamadı");
                }
            }
            else if (s === "4") {
                const aramaId = await soru("Nesne ID (ilk 8 karakter): ");
                const nesne = envanter.find(x => x.id.startsWith(aramaId.trim()));

                if (nesne) {
                    // Type Checking: Symbol marker ile IKritik kontrolü
                    if (nesne[IKritik]) {
                        nesne.acilDurumSogutmasi();
                    } else {
                        console.log(" Bu nesne soğutulamaz! (IKritik değil)");
                    }
                } else {
                    console.log(" Nesne bulunamadı");
                }
            }
            else if (s === "5") {
                console.log("Sistem kapatılıyor");
                rl.close();
                process.exit(0);
            }
        }
    }
    catch (error) {
        if (error instanceof KuantumCokusuException) {
            console.log("\n" + "=".repeat(50));
            console.log(" SİSTEM ÇÖKTÜ! TAHLİYE BAŞLATILIYOR");
            console.log(error.message);
            console.log("=".repeat(50));
        } else {
            console.error("Beklenmeyen hata:", error);
        }
        rl.close();
        process.exit(1);
    }
})();